src/InjectHostDevBufferCopies.cpp

```diff
@@ -504,19 +513,76 @@ class InjectBufferCopies : public IRMutator {
             stmt = Allocate::make(op->name, op->type, op->extents, const_false(), op->body);
+        } else if (buf_info.on_single_device &&
+                   buf_info.dev_touched) {
+            debug(4) << "Making combined host/device alloc for " << op->name << "\n";
+            Stmt inner_body = op->body;
+            std::vector<const LetStmt *> body_lets;
+            // Find LetStmt seting up buffer Variable for op->name as it will
+            // now go outside.
+            const LetStmt *buffer_create_let = nullptr;
+            while (const LetStmt *inner_let = inner_body.as<LetStmt>()) {
+                inner_body = inner_let->body;
+                if (inner_let->name == op->name + ".buffer") {
+                    buffer_create_let = inner_let;
+                    break;
+                }
+                body_lets.push_back(inner_let);
+            }
+
+            // Rebuild any wrapped lets without the one for the create_buffer_t.
+            for (size_t i = body_lets.size(); i > 0; i--) {
+                inner_body = LetStmt::make(body_lets[i - 1]->name, body_lets[i - 1]->value, inner_body);
+            }
+
+            Stmt combined_malloc = make_dev_malloc(op->name, buf_info.device_first_touched, true);
+
+            // Create a new Allocation scope inside the buffer
+            // creation, use the host pointer as the allocation and
+            // set the destructor to a nop.  (The Allocation
+            // destructor cannot be used as it takes the host pointer
+            // as it's argument and we need the complete buffer_t.  it
+            // would be possible to keep a map between host pointers
+            // and dev ones to facilitate this, but it seems better to
+            // just register a destructor with the buffer creation.)
+            inner_body = Allocate::make(op->name, op->type, op->extents, op->condition, inner_body,
+                                        Call::make(Handle(), Call::extract_buffer_host,
+                                                   { Variable::make(type_of<struct buffer_t *>(), op->name + ".buffer") },
+                                                   Call::Intrinsic),
+                                        "halide_device_host_nop_free"); // TODO: really should not have to introduce this routine to get a nop free
+            // Wrap combined malloc around Allocate.
+            inner_body = Block::make(combined_malloc, inner_body);
+
+            // Rewrite original create_buffer_t call and wrap it around the combined malloc.
+            std::vector<Expr> create_buffer_args;
+            if (buffer_create_let != nullptr) {
+                const Call *possible_create_buffer = buffer_create_let->value.as<Call>();
+                if (possible_create_buffer != nullptr &&
+                    possible_create_buffer->is_intrinsic(Call::create_buffer_t)) {
+                    create_buffer_args = possible_create_buffer->args;
```

Comment:
I nearly suggested that but was worrying about cases like:

Allocate foo
let x = address_of(foo[0])
let y = create_buffer( (something that depends on x) )

It's still technically not a safe IR mutation, though I'm not sure that case could ever occur. If it did, I'm also not sure what we could do other than throw an error. Making allocations and associated buffers really needs to be one atomic thing, but we should probably kick that can down the road for now.